### PR TITLE
P6 protocol update

### PR DESCRIPTION
**Pull request by MilkywayPirate**
## Purpose

Address https://github.com/Concordium/concordium-node/issues/856 by implementing the protocol upgrade P5->P6.

## Changes

- Implement the migrations for the block state.
    - Note that for this protocol update we are updating the seed state as part of the migration `migrateSeedState`.
- Introduce a function `migrateSkovFromConsensusV0` which performs the migration from a `ConsensusV0` skov to a `ConsensusV1` skov. This is being called from the `MultiversionRunner` as part of `checkForProtocolUpdate`.
- Implemented the necessary code in `checkForProtocolUpdate` which performs the migration and starts the new protocol.
- Introduced an `Update` from `P5` for `ProtocolP6`.
  Note that `ProtocolP6` is currently using a dummy hash value and hence must be updated when the protocol update specification is ready.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Review by eb-concordium — COMMENTED** [review]
Looks good to me, modulo a comment regarding the migration of delegators when updating P5->P6.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/Bakers.hs [hunk 1]
```diff
@@ -293,6 +294,8 @@ migratePersistentActiveDelegators StateMigrationParametersP4ToP5{} =
     \PersistentActiveDelegatorsV1{..} -> do
         newDelegators <- Trie.migrateTrieN True return adDelegators
         return PersistentActiveDelegatorsV1{adDelegators = newDelegators, ..}
+migratePersistentActiveDelegators StateMigrationParametersP5ToP6{} = \case
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/Bakers.hs:297]: Yes, you're right!

**Review by eb-concordium — APPROVED** [review]

File: concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs [hunk 2]
```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- FIXME: This is currently a dummy update, and the details, including the update hash, need to be resolved before release.
+-- TODO: Consider moving the state changes that are made here into the state migration. [This cannot be done after release.]
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs:7]: What does this TODO refer to?

Which state changes are done here?

File: concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+-- * 'genesisStateHash' and 'genesisNewState' are the hash and (V0) serialized state of the
+--   new genesis block, which are derived from the block state of the last finalized block of
+--   the previous chain by applying the following changes:
+--
+--     * The 'SeedState' is updated with:
+--
+--         * 'epochLength' is unchanged;
+--         * 'epoch' is @0@;
+--         * 'currentLeadershipElectionNonce' is the SHA256 hash of (@"Regenesis" <> encode (updatedNonce oldSeedState)@); and
+--         * 'updatedNonce' is the same as 'currentLeadershipElectionNonce'.
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs:42]: I cannot see this being done in the code below.

File: concordium-consensus/src/Concordium/KonsensusV1/Types.hs [hunk 4]
```diff
@@ -1242,6 +1244,19 @@ data GenesisMetadata = GenesisMetadata
     }
     deriving (Eq, Show)
 
+-- |Extract the genesis configuration from the regenesis data.
+regenesisMetadata :: (IsProtocolVersion pv, IsConsensusV1 pv) => StateHash -> Regenesis pv -> GenesisMetadata
+regenesisMetadata sh regenData =
+    GenesisMetadata
+        { gmParameters = regenesisCoreParametersV1 regenData,
+          -- \|Hash of the genesis block.
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/KonsensusV1/Types.hs:1252]: Hmm.. Removed. I do not recall putting those there. Fixed.

File: concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs [hunk 5]
```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- FIXME: This is currently a dummy update, and the details, including the update hash, need to be resolved before release.
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs:6]: I added it to the fixme of the hash and at the top of the file.

File: concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs (hunk 2 above)
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs:7]: copy & paste error.

File: concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+--
+--         * 'epochLength' is unchanged;
+--         * 'epoch' is @0@;
+--         * 'currentLeadershipElectionNonce' is the SHA256 hash of (@"Regenesis" <> encode (updatedNonce oldSeedState)@); and
+--         * 'updatedNonce' is the same as 'currentLeadershipElectionNonce'.
+--
+--     * The 'Updates' are updated with:
+--
+--         * the current protocol update is set to 'Nothing'; and
+--         * the protocol update queue is emptied.
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs:47]: copy and paste error.

File: concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs (hunk 3 above)
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/ProtocolUpdate/P5/ProtocolP6.hs:42]: copy and pase err

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/Updates.hs [hunk 7]
```diff
@@ -716,7 +740,12 @@ migrateUpdates migration Updates{..} = do
     newParameters <- migrateHashedBufferedRef (return . StoreSerialized . migrateChainParameters migration . unStoreSerialized) currentParameters
     newCurrentProtocolUpdate <- case currentProtocolUpdate of
         Null -> return Null
-        Some c -> Some <$!> migrateHashedBufferedRefKeepHash c
+        -- For P5->P6 we clear the currentProtocolUpdate as part of the migration.
+        -- For older versions the @currentProtocolUpdate@ is cleared before i.e. as
+        -- part of 'updateRegenesis'.
+        Some c -> case (protocolVersion @pv) of
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/Updates.hs:746]: You could probably just clear it always. But I haven't checked if we actually rely on it anywhere later. I don't think we do, but this is also fine.

**Review by abizjak — APPROVED** [review]

**Review by td202 — CHANGES_REQUESTED** [review]
Looks good. A few issues need to be addressed before merge.

**Pull request merged**